### PR TITLE
fby35: cl: Support MCTP/PLDM to IPMB bridge command

### PR DESCRIPTION
**Pull request by DelphineChiu**
Summary:
- Support BIC to send MCTP/PLDM request to BMC.
- Support bridge command from BMC to ME/1OU BIC (BMC -> MCTP/PLDM -> BIC -> IPMB -> KCS/ME/1OU BIC).
- Support KCS/ME request to BMC (KCS/ME -> IPMB -> BIC -> MCTP/PLDM -> BMC).
- Support pldm instance id field.
- Fix error message when sending GPIO interrupt to BMC.

Dependency: #809 

Test plan:
- Build code: Pass
- Using bridge command: Pass
- Host power on: Pass
- Using IPMI command on host: Pass
- Fix send GPIO interrupt error: Pass

Log:
1. Get 1OU BIC version.
root@bmc-oob:~# fw-util slot1 --version 1ou_bic
1OU Bridge-IC Version: oby35-vf-v2022.47.01

2. Get ME device ID.
root@bmc-oob:~# me-util slot1 --get_dev_id
Device ID:                 50
Device Revision:           01
Firmware Revision:         06 03
IPMI Version:              02
Additional Device Support: 21
Manufacturer ID:           57 01 00
Product ID:                18 0b
Aux Firmware Revision:     06 27 30

3. Get BMC device ID from Host.
[root@FBK8_221107 ~]# ipmitool raw 0x6 0x1
 20 81 01 03 02 bf 15 a0 00 46 31 00 00 00 00

4. Fix send GPIO interrupt error.
- Before fix
uart:~$ Send gpio interrupt to BMC, gpio number(39) status(0)
Failed to send GPIO interrupt event to BMC, gpio number(39) status(1)
Send gpio interrupt to BMC, gpio number(32) status(0)
Failed to send GPIO interrupt event to BMC, gpio number(32) status(1)

- After fix
[00:00:45.510,000] <inf> plat_isr: Send gpio interrupt to BMC, gpio number(32) status(1)
[00:00:46.170,000] <inf> plat_isr: Send gpio interrupt to BMC, gpio number(39) status(1)

### CODE REVIEW

**Comment by DelphineChiu** [comment]
Dependency: #809

File: common/service/host/kcs.c [hunk 1]
```diff
@@ -149,9 +150,44 @@ void kcs_read(void *arvg0, void *arvg1, void *arvg2)
...
+				int ret = 0;
+
+				// Send request to MCTP/PLDM thread to ask BMC
+				ret = pldm_send_ipmi_request(&bridge_msg);
+				if (ret < 0) {
+					LOG_ERR("kcs_read_task send to BMC fail");
+				}
+
+				uint8_t *kcs_buff;
+				kcs_buff = malloc(KCS_BUFF_SIZE * sizeof(uint8_t));
```
Comment (garnermic) [inline common/service/host/kcs.c:165]: Do we want to allocate a static buffer, or should we allocate 3+bridge_msg.data_len?

File: common/service/host/kcs.c [hunk 2]
```diff
@@ -149,9 +150,44 @@ void kcs_read(void *arvg0, void *arvg1, void *arvg2)
...
+					LOG_ERR("Memory allocation failed");
+					continue;
+				}
+
+				kcs_buff[0] = (bridge_msg.netfn | BIT(0)) << 2;
+				kcs_buff[1] = bridge_msg.cmd;
+				kcs_buff[2] = bridge_msg.completion_code;
+
+				// Write MCTP/PLDM response to KCS
+				if (bridge_msg.data_len > 0) {
```
Comment (garnermic) [inline common/service/host/kcs.c:176]: Ideally we would add an iovec-style kcs_write routine in order to reduce the amount of memcpy and buffers we need to allocate.

**Review by garnermic — CHANGES_REQUESTED** [review]
Changes requested based on Meta review.

**Comment by SaraLin-wiwynn** [comment]
Hi @garnermic
Thanks for your comments.
Based on your suggestions to modify the code, rebase main and force update this branch.
Please help review it again, thank you.

File: common/service/host/kcs.c (hunk 1 above)
Comment (SaraLin-wiwynn) [reply common/service/host/kcs.c]: Change to allocate 3 + bridge_msg.data_len, thank you.

File: common/service/host/kcs.c (hunk 2 above)
Comment (SaraLin-wiwynn) [reply common/service/host/kcs.c]: Could you explain iovec-style routine in more detail? 
We're not sure we understand it.
And could we update this next commit?
Thank you.

Comment (garnermic) [reply common/service/host/kcs.c]: Here is a reference on vectored I/O support: https://dius-au.medium.com/good-design-from-the-ground-up-how-i-o-library-design-can-make-or-break-iot-solutions-49a66bc6c5b2 which Patrick is referencing on how to relieve memory pressure/usage.

**Comment by SaraLin-wiwynn** [comment]
Modify the code to fix the build failed problem and force update this branch.
Please help review it again, thank you.

**Pull request closed without merging**